### PR TITLE
HDDS-5865. Make read retry interval and attempts in BlockInputStream configurable

### PR DESCRIPTION
**Pull request by SaketaChalamchala**
## What changes were proposed in this pull request?
Retry policy when reading keys is hardcoded. Made the read retry max attempts and retry interval configurable in client configuration ozone.client.max.read.retries and ozone.client.max.read.retries. Modified InputStream classes to accept client configuration so that read retry policy can be created wherever required.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5865

## How was this patch tested?

Existing UT and integration tests should run as is.

### CODE REVIEW

**Comment by SaketaChalamchala** [comment]
@adoroszlai and @ashishkumar50 can you please take a look

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 1]
```diff
@@ -111,26 +112,32 @@ public class BlockInputStream extends BlockExtendedInputStream {
...
       Token<OzoneBlockTokenIdentifier> token, boolean verifyChecksum,
       XceiverClientFactory xceiverClientFactory,
-      Function<BlockID, BlockLocationInfo> refreshFunction) {
+      Function<BlockID, BlockLocationInfo> refreshFunction,
+      OzoneClientConfig config) {
     this.blockID = blockId;
     this.length = blockLen;
     setPipeline(pipeline);
     tokenRef.set(token);
     this.verifyChecksum = verifyChecksum;
```
[the remark this answers is not in the document]
Comment (SaketaChalamchala) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java]: Done

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 2]
```diff
@@ -111,26 +112,32 @@ public class BlockInputStream extends BlockExtendedInputStream {
 
   private final Function<BlockID, BlockLocationInfo> refreshFunction;
 
+  @SuppressWarnings("checkstyle:ParameterNumber")
   public BlockInputStream(BlockID blockId, long blockLen, Pipeline pipeline,
       Token<OzoneBlockTokenIdentifier> token, boolean verifyChecksum,
       XceiverClientFactory xceiverClientFactory,
-      Function<BlockID, BlockLocationInfo> refreshFunction) {
+      Function<BlockID, BlockLocationInfo> refreshFunction,
+      OzoneClientConfig config) {
```
[the remark this answers is not in the document]
Comment (SaketaChalamchala) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java]: Thanks for the review @adoroszlai . I made the requested changes. Could you please review them?

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/OzoneClientConfig.java [hunk 3]
```diff
@@ -144,6 +144,23 @@ public enum ChecksumCombineMode {
...
+      tags = ConfigTag.CLIENT)
+  private int maxReadRetryCount = 3;
+
+  @Config(key = "read.retry.interval",
+      defaultValue = "1",
+      description =
+          "Indicates the time duration in seconds a client will wait "
+              + "before retrying a read key request on encountering "
+              + "a connectivity excepetion from Datanodes . "
+              + "By default the interval is 1millisecond",
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/OzoneClientConfig.java:160]:
```suggestion
              + "By default the interval is 1 second",
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @SaketaChalamchala for updating the patch, LGTM.  (Fixed a typo in the config description.  Sorry I missed it earlier.)

**Review by ashishkumar50 — COMMENTED** [review]
@SaketaChalamchala Thanks for updating the patch, LGTM.

**Comment by adoroszlai** [comment]
Thanks @SaketaChalamchala for the patch, @ashishkumar50 for the review.

**Pull request merged**